File: src/screens/HomeScreen.js
import React, {useEffect} from 'react';
import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
  RefreshControl,
  StatusBar,
} from 'react-native';
import {COLORS, FONTS, SIZES, SPACING} from '../constants';
import {useAuth} from '../context/AuthContext';
import {useEnergy} from '../context/EnergyContext';
import EnergyCard from '../components/EnergyCard';
import DeviceCard from '../components/DeviceCard';
import QuickActionCard from '../components/QuickActionCard';

const HomeScreen = ({navigation}) => {
  const {user} = useAuth();
  const {
    currentEnergyData,
    devices,
    isLoading,
    fetchEnergyData,
    fetchDevices,
  } = useEnergy();

  useEffect(() => {
    loadData();
  }, []);

  const loadData = async () => {
    await Promise.all([fetchEnergyData(), fetchDevices()]);
  };

  const onRefresh = () => {
    loadData();
  };

  const renderHeader = () => (
    <View style={styles.header}>
      <View style={styles.profileSection}>
        <View style={styles.avatar}>
          <Text style={styles.avatarText}>
            {user?.name?.charAt(0)?.toUpperCase() || 'U'}
          </Text>
        </View>
        <View style={styles.welcomeSection}>
          <Text style={styles.welcomeText}>
            Welcome back, {user?.name || 'User'}!
          </Text>
          <Text style={styles.subtitle}>Here's your energy overview</Text>
        </View>
      </View>
      <TouchableOpacity style={styles.notificationButton}>
        <Text style={styles.notificationIcon}>🔔</Text>
      </TouchableOpacity>
    </View>
  );

  const renderWelcomeCard = () => (
    <View style={styles.welcomeCard}>
      <View style={styles.welcomeContent}>
        <Text style={styles.welcomeCardTitle}>Smart Energy Management</Text>
        <Text style={styles.welcomeCardSubtitle}>
          Monitor and optimize your energy consumption
        </Text>
      </View>
      <Text style={styles.welcomeIcon}>⚡</Text>
    </View>
  );

  const renderEnergyOverview = () => (
    <View style={styles.section}>
      <Text style={styles.sectionTitle}>Energy Overview</Text>
      <View style={styles.energyCards}>
        <EnergyCard
          title="Current Usage"
          value={`${currentEnergyData?.consumption?.toFixed(1) || '0.0'} kWh`}
          icon="⚡"
          color={COLORS.primary}
        />
        <EnergyCard
          title="Cost Today"
          value={`$${currentEnergyData?.cost?.toFixed(2) || '0.00'}`}
          icon="💰"
          color={COLORS.success}
        />
      </View>
    </View>
  );

  const renderQuickActions = () => (
    <View style={styles.section}>
      <Text style={styles.sectionTitle}>Quick Actions</Text>
      <View style={styles.quickActionsGrid}>
        <QuickActionCard
          title="Upload Bill"
          icon="📄"
          color={COLORS.primary}
          onPress={() => navigation.navigate('UploadBill')}
        />
        <QuickActionCard
          title="Add Device"
          icon="📱"
          color={COLORS.accent}
          onPress={() => navigation.navigate('DeviceSetup')}
        />
        <QuickActionCard
          title="Better Tariffs"
          icon="💰"
          color={COLORS.success}
          onPress={() => navigation.navigate('BetterTariffs')}
        />
        <QuickActionCard
          title="Energy Report"
          icon="📊"
          color={COLORS.warning}
          onPress={() => navigation.navigate('EnergyReport')}
        />
      </View>
    </View>
  );

  const renderDevices = () => (
    <View style={styles.section}>
      <View style={styles.sectionHeader}>
        <Text style={styles.sectionTitle}>Your Devices</Text>
        <TouchableOpacity onPress={() => navigation.navigate('Devices')}>
          <Text style={styles.viewAllText}>View All</Text>
        </TouchableOpacity>
      </View>
      {devices.length === 0 ? (
        <View style={styles.emptyDevices}>
          <Text style={styles.emptyIcon}>📱</Text>
          <Text style={styles.emptyTitle}>No devices connected</Text>
          <Text style={styles.emptySubtitle}>
            Add your first smart device to start monitoring
          </Text>
        </View>
      ) : (
        devices.slice(0, 3).map(device => (
          <DeviceCard key={device.id} device={device} />
        ))
      )}
    </View>
  );

  const renderRecentActivity = () => (
    <View style={styles.section}>
      <Text style={styles.sectionTitle}>Recent Activity</Text>
      <View style={styles.activityCard}>
        <View style={styles.activityItem}>
          <View style={[styles.activityIcon, {backgroundColor: COLORS.warning + '20'}]}>
            <Text style={styles.activityIconText}>⚡</Text>
          </View>
          <View style={styles.activityContent}>
            <Text style={styles.activityTitle}>High energy usage detected</Text>
            <Text style={styles.activitySubtitle}>
              Kitchen appliances are using more power than usual
            </Text>
          </View>
          <Text style={styles.activityTime}>2h ago</Text>
        </View>
        <View style={styles.activityDivider} />
        <View style={styles.activityItem}>
          <View style={[styles.activityIcon, {backgroundColor: COLORS.success + '20'}]}>
            <Text style={styles.activityIconText}>💰</Text>
          </View>
          <View style={styles.activityContent}>
            <Text style={styles.activityTitle}>New tariff available</Text>
            <Text style={styles.activitySubtitle}>Save 15% on your monthly bill</Text>
          </View>
          <Text style={styles.activityTime}>1d ago</Text>
        </View>
        <View style={styles.activityDivider} />
        <View style={styles.activityItem}>
          <View style={[styles.activityIcon, {backgroundColor: COLORS.primary + '20'}]}>
            <Text style={styles.activityIconText}>📱</Text>
          </View>
          <View style={styles.activityContent}>
            <Text style={styles.activityTitle}>Device connected</Text>
            <Text style={styles.activitySubtitle}>Smart plug added to living room</Text>
          </View>
          <Text style={styles.activityTime}>2d ago</Text>
        </View>
      </View>
    </View>
  );

  return (
    <View style={styles.container}>
      <StatusBar
        barStyle="dark-content"
        backgroundColor={COLORS.white}
        translucent={false}
      />
      
      <ScrollView
        style={styles.scrollView}
        contentContainerStyle={styles.scrollContent}
        refreshControl={
          <RefreshControl refreshing={isLoading} onRefresh={onRefresh} />
        }
        showsVerticalScrollIndicator={false}>
        {renderHeader()}
        {renderWelcomeCard()}
        {renderEnergyOverview()}
        {renderQuickActions()}
        {renderDevices()}
        {renderRecentActivity()}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: COLORS.background,
  },
  scrollView: {
    flex: 1,
  },
  scrollContent: {
    paddingBottom: SPACING.xl,
  },
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: SPACING.xl,
    paddingVertical: SPACING.lg,
    backgroundColor: COLORS.white,
    borderBottomWidth: 1,
    borderBottomColor: COLORS.border,
  },
  profileSection: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  avatar: {
    width: 40,
    height: 40,
    borderRadius: 20,
    backgroundColor: COLORS.primary + '20',
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: SPACING.md,
  },
  avatarText: {
    fontSize: 18,
    fontFamily: FONTS.bold,
    color: COLORS.primary,
  },
  welcomeSection: {
    flex: 1,
  },
  welcomeText: {
    fontSize: 16,
    fontFamily: FONTS.semiBold,
    color: COLORS.textPrimary,
  },
  subtitle: {
    fontSize: 14,
    fontFamily: FONTS.regular,
    color: COLORS.textSecondary,
  },
  notificationButton: {
    padding: SPACING.sm,
  },
  notificationIcon: {
    fontSize: 24,
  },
  welcomeCard: {
    margin: SPACING.xl,
    padding: SPACING.xl,
    backgroundColor: COLORS.primary,
    borderRadius: 16,
    flexDirection: 'row',
    alignItems: 'center',
  },
  welcomeContent: {
    flex: 1,
  },
  welcomeCardTitle: {
    fontSize: 18,
    fontFamily: FONTS.bold,
    color: COLORS.white,
    marginBottom: SPACING.sm,
  },
  welcomeCardSubtitle: {
    fontSize: 14,
    fontFamily: FONTS.regular,
    color: COLORS.white,
    opacity: 0.9,
  },
  welcomeIcon: {
    fontSize: 48,
  },
  section: {
    paddingHorizontal: SPACING.xl,
    marginBottom: SPACING.xl,
  },
  sectionHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: SPACING.lg,
  },
  sectionTitle: {
    fontSize: 18,
    fontFamily: FONTS.bold,
    color: COLORS.textPrimary,
    marginBottom: SPACING.lg,
  },
  viewAllText: {
    fontSize: 14,
    fontFamily: FONTS.medium,
    color: COLORS.primary,
  },
  energyCards: {
    flexDirection: 'row',
    gap: SPACING.md,
  },
  quickActionsGrid: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    gap: SPACING.md,
  },
  emptyDevices: {
    alignItems: 'center',
    padding: SPACING.xl,
    backgroundColor: COLORS.white,
    borderRadius: 12,
    borderWidth: 1,
    borderColor: COLORS.border,
  },
  emptyIcon: {
    fontSize: 48,
    marginBottom: SPACING.md,
  },
  emptyTitle: {
    fontSize: 16,
    fontFamily: FONTS.semiBold,
    color: COLORS.textPrimary,
    marginBottom: SPACING.sm,
  },
  emptySubtitle: {
    fontSize: 14,
    fontFamily: FONTS.regular,
    color: COLORS.textSecondary,
    textAlign: 'center',
  },
  activityCard: {
    backgroundColor: COLORS.white,
    borderRadius: 12,
    borderWidth: 1,
    borderColor: COLORS.border,
    padding: SPACING.lg,
  },
  activityItem: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  activityIcon: {
    width: 40,
    height: 40,
    borderRadius: 8,
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: SPACING.md,
  },
  activityIconText: {
    fontSize: 20,
  },
  activityContent: {
    flex: 1,
  },
  activityTitle: {
    fontSize: 14,
    fontFamily: FONTS.semiBold,
    color: COLORS.textPrimary,
    marginBottom: SPACING.xs,
  },
  activitySubtitle: {
    fontSize: 12,
    fontFamily: FONTS.regular,
    color: COLORS.textSecondary,
  },
  activityTime: {
    fontSize: 12,
    fontFamily: FONTS.regular,
    color: COLORS.textTertiary,
  },
  activityDivider: {
    height: 1,
    backgroundColor: COLORS.border,
    marginVertical: SPACING.md,
  },
});

export default HomeScreen; 

File: lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';
import 'package:provider/provider.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:e_snapp_mobile/providers/auth_provider.dart';
import 'package:e_snapp_mobile/providers/energy_provider.dart';
import 'package:e_snapp_mobile/utils/constants.dart';
import 'package:e_snapp_mobile/widgets/bottom_navigation.dart';
import 'package:e_snapp_mobile/widgets/energy_card.dart';
import 'package:e_snapp_mobile/widgets/device_card.dart';
import 'package:e_snapp_mobile/widgets/quick_action_card.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  @override
  void initState() {
    super.initState();
    _loadData();
  }

  Future<void> _loadData() async {
    final energyProvider = Provider.of<EnergyProvider>(context, listen: false);
    await energyProvider.fetchEnergyData();
    await energyProvider.fetchDevices();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AppColors.background,
      body: SafeArea(
        child: Column(
          children: [
            // App Bar
            _buildAppBar(),
            
            // Content
            Expanded(
              child: RefreshIndicator(
                onRefresh: _loadData,
                child: SingleChildScrollView(
                  padding: EdgeInsets.all(16.w),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      // Welcome Section
                      _buildWelcomeSection(),
                      SizedBox(height: 24.h),
                      
                      // Energy Overview
                      _buildEnergyOverview(),
                      SizedBox(height: 24.h),
                      
                      // Quick Actions
                      _buildQuickActions(),
                      SizedBox(height: 24.h),
                      
                      // Devices Section
                      _buildDevicesSection(),
                      SizedBox(height: 24.h),
                      
                      // Recent Activity
                      _buildRecentActivity(),
                    ],
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
      bottomNavigationBar: const BottomNavigation(),
    );
  }

  Widget _buildAppBar() {
    return Container(
      padding: EdgeInsets.all(16.w),
      decoration: BoxDecoration(
        color: Colors.white,
        boxShadow: [
          BoxShadow(
            color: Colors.black.withOpacity(0.05),
            blurRadius: 10,
            offset: const Offset(0, 2),
          ),
        ],
      ),
      child: Row(
        children: [
          // Profile Avatar
          Consumer<AuthProvider>(
            builder: (context, authProvider, child) {
              return CircleAvatar(
                radius: 20.w,
                backgroundColor: AppColors.primary.withOpacity(0.1),
                child: Text(
                  authProvider.user?.name.substring(0, 1).toUpperCase() ?? 'U',
                  style: TextStyle(
                    fontSize: 18.sp,
                    fontWeight: FontWeight.bold,
                    color: AppColors.primary,
                  ),
                ),
              );
            },
          ),
          SizedBox(width: 12.w),
          
          // Welcome Text
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Consumer<AuthProvider>(
                  builder: (context, authProvider, child) {
                    return Text(
                      'Welcome back, ${authProvider.user?.name ?? 'User'}!',
                      style: TextStyle(
                        fontSize: 16.sp,
                        fontWeight: FontWeight.w600,
                        color: AppColors.textPrimary,
                      ),
                    );
                  },
                ),
                Text(
                  'Here\'s your energy overview',
                  style: TextStyle(
                    fontSize: 14.sp,
                    color: AppColors.textSecondary,
                  ),
                ),
              ],
            ),
          ),
          
          // Notifications
          IconButton(
            onPressed: () {
              // TODO: Navigate to notifications
            },
            icon: Icon(
              Icons.notifications_outlined,
              size: 24.w,
              color: AppColors.textSecondary,
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildWelcomeSection() {
    return Container(
      padding: EdgeInsets.all(20.w),
      decoration: BoxDecoration(
        gradient: LinearGradient(
          colors: [AppColors.primary, AppColors.primary.withOpacity(0.8)],
          begin: Alignment.topLeft,
          end: Alignment.bottomRight,
        ),
        borderRadius: BorderRadius.circular(16),
      ),
      child: Row(
        children: [
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  'Smart Energy Management',
                  style: TextStyle(
                    fontSize: 18.sp,
                    fontWeight: FontWeight.bold,
                    color: Colors.white,
                  ),
                ),
                SizedBox(height: 8.h),
                Text(
                  'Monitor and optimize your energy consumption',
                  style: TextStyle(
                    fontSize: 14.sp,
                    color: Colors.white.withOpacity(0.9),
                  ),
                ),
              ],
            ),
          ),
          Icon(
            Icons.flash_on,
            size: 48.w,
            color: Colors.white,
          ),
        ],
      ),
    );
  }

  Widget _buildEnergyOverview() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          'Energy Overview',
          style: TextStyle(
            fontSize: 18.sp,
            fontWeight: FontWeight.bold,
            color: AppColors.textPrimary,
          ),
        ),
        SizedBox(height: 16.h),
        Consumer<EnergyProvider>(
          builder: (context, energyProvider, child) {
            if (energyProvider.isLoading) {
              return const Center(child: CircularProgressIndicator());
            }
            
            final currentData = energyProvider.currentEnergyData;
            if (currentData == null) {
              return const Center(
                child: Text('No energy data available'),
              );
            }
            
            return Row(
              children: [
                Expanded(
                  child: EnergyCard(
                    title: 'Current Usage',
                    value: '${currentData.consumption.toStringAsFixed(1)} kWh',
                    icon: Icons.electric_bolt,
                    color: AppColors.primary,
                  ),
                ),
                SizedBox(width: 12.w),
                Expanded(
                  child: EnergyCard(
                    title: 'Cost Today',
                    value: '\$${currentData.cost.toStringAsFixed(2)}',
                    icon: Icons.attach_money,
                    color: AppColors.success,
                  ),
                ),
              ],
            );
          },
        ),
      ],
    );
  }

  Widget _buildQuickActions() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          'Quick Actions',
          style: TextStyle(
            fontSize: 18.sp,
            fontWeight: FontWeight.bold,
            color: AppColors.textPrimary,
          ),
        ),
        SizedBox(height: 16.h),
        Row(
          children: [
            Expanded(
              child: QuickActionCard(
                title: 'Upload Bill',
                icon: Icons.upload_file,
                color: AppColors.primary,
                onTap: () {
                  // TODO: Navigate to upload bill screen
                },
              ),
            ),
            SizedBox(width: 12.w),
            Expanded(
              child: QuickActionCard(
                title: 'Add Device',
                icon: Icons.add_circle_outline,
                color: AppColors.accent,
                onTap: () {
                  // TODO: Navigate to add device screen
                },
              ),
            ),
          ],
        ),
        SizedBox(height: 12.h),
        Row(
          children: [
            Expanded(
              child: QuickActionCard(
                title: 'Better Tariffs',
                icon: Icons.savings,
                color: AppColors.success,
                onTap: () {
                  // TODO: Navigate to better tariffs screen
                },
              ),
            ),
            SizedBox(width: 12.w),
            Expanded(
              child: QuickActionCard(
                title: 'Energy Report',
                icon: Icons.analytics,
                color: AppColors.warning,
                onTap: () {
                  // TODO: Navigate to energy report screen
                },
              ),
            ),
          ],
        ),
      ],
    );
  }

  Widget _buildDevicesSection() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(
              'Your Devices',
              style: TextStyle(
                fontSize: 18.sp,
                fontWeight: FontWeight.bold,
                color: AppColors.textPrimary,
              ),
            ),
            TextButton(
              onPressed: () {
                // TODO: Navigate to devices screen
              },
              child: Text(
                'View All',
                style: TextStyle(
                  fontSize: 14.sp,
                  color: AppColors.primary,
                ),
              ),
            ),
          ],
        ),
        SizedBox(height: 16.h),
        Consumer<EnergyProvider>(
          builder: (context, energyProvider, child) {
            if (energyProvider.isLoading) {
              return const Center(child: CircularProgressIndicator());
            }
            
            final devices = energyProvider.devices;
            if (devices.isEmpty) {
              return Container(
                padding: EdgeInsets.all(20.w),
                decoration: BoxDecoration(
                  color: Colors.white,
                  borderRadius: BorderRadius.circular(12),
                  border: Border.all(color: AppColors.border),
                ),
                child: Column(
                  children: [
                    Icon(
                      Icons.devices_outlined,
                      size: 48.w,
                      color: AppColors.textSecondary,
                    ),
                    SizedBox(height: 12.h),
                    Text(
                      'No devices connected',
                      style: TextStyle(
                        fontSize: 16.sp,
                        fontWeight: FontWeight.w600,
                        color: AppColors.textPrimary,
                      ),
                    ),
                    SizedBox(height: 8.h),
                    Text(
                      'Add your first smart device to start monitoring',
                      style: TextStyle(
                        fontSize: 14.sp,
                        color: AppColors.textSecondary,
                        textAlign: TextAlign.center,
                      ),
                    ),
                  ],
                ),
              );
            }
            
            return Column(
              children: devices.take(3).map((device) => DeviceCard(device: device)).toList(),
            );
          },
        ),
      ],
    );
  }

  Widget _buildRecentActivity() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          'Recent Activity',
          style: TextStyle(
            fontSize: 18.sp,
            fontWeight: FontWeight.bold,
            color: AppColors.textPrimary,
          ),
        ),
        SizedBox(height: 16.h),
        Container(
          padding: EdgeInsets.all(16.w),
          decoration: BoxDecoration(
            color: Colors.white,
            borderRadius: BorderRadius.circular(12),
            border: Border.all(color: AppColors.border),
          ),
          child: Column(
            children: [
              _buildActivityItem(
                icon: Icons.electric_bolt,
                title: 'High energy usage detected',
                subtitle: 'Kitchen appliances are using more power than usual',
                time: '2 hours ago',
                color: AppColors.warning,
              ),
              Divider(color: AppColors.border),
              _buildActivityItem(
                icon: Icons.savings,
                title: 'New tariff available',
                subtitle: 'Save 15% on your monthly bill',
                time: '1 day ago',
                color: AppColors.success,
              ),
              Divider(color: AppColors.border),
              _buildActivityItem(
                icon: Icons.devices,
                title: 'Device connected',
                subtitle: 'Smart plug added to living room',
                time: '2 days ago',
                color: AppColors.primary,
              ),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildActivityItem({
    required IconData icon,
    required String title,
    required String subtitle,
    required String time,
    required Color color,
  }) {
    return Row(
      children: [
        Container(
          width: 40.w,
          height: 40.w,
          decoration: BoxDecoration(
            color: color.withOpacity(0.1),
            borderRadius: BorderRadius.circular(8),
          ),
          child: Icon(
            icon,
            size: 20.w,
            color: color,
          ),
        ),
        SizedBox(width: 12.w),
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                title,
                style: TextStyle(
                  fontSize: 14.sp,
                  fontWeight: FontWeight.w600,
                  color: AppColors.textPrimary,
                ),
              ),
              SizedBox(height: 4.h),
              Text(
                subtitle,
                style: TextStyle(
                  fontSize: 12.sp,
                  color: AppColors.textSecondary,
                ),
              ),
            ],
          ),
        ),
        Text(
          time,
          style: TextStyle(
            fontSize: 12.sp,
            color: AppColors.textTertiary,
          ),
        ),
      ],
    );
  }
} 